source/framefmt.c

```diff
@@ -235,16 +252,23 @@ int aws_cryptosdk_serialize_frame(
     size_t plaintext_size,
     struct aws_byte_buf *ciphertext_buf,
     const struct aws_cryptosdk_alg_properties *alg_props) {
+    AWS_PRECONDITION(aws_cryptosdk_alg_properties_is_valid(alg_props));
     struct aws_cryptosdk_framestate state;
 
+    // The plaintext_size should be bound to prevent arithmetic
+    // overflows due to addition
+    if (plaintext_size > MAX_PLAINTEXT_SIZE) {
+        return aws_raise_error(AWS_CRYPTOSDK_ERR_LIMIT_EXCEEDED);
+    }
+    
     // We assume that the max frame size is equal to the plaintext size. This
     // lets us avoid having to pass in a redundant argument, avoids needing to
     // take a branch in serde_framed, and does not impact the serialized
     // output.
     state.max_frame_size = plaintext_size;
     state.plaintext_size = plaintext_size;
     // Currently all supported algorithms have plaintext = ciphertext size
-    state.ciphertext_size = plaintext_size;
+    state.ciphertext_size = 0;
```

Comment:
I modified this so that the assertion [in line 45 of the harness](https://github.com/aws/aws-encryption-sdk-c/pull/378/files#diff-ac6b8365f754b91ff670c36d2435058dR45) passes. Is that assertion reasonable?

Reply:
I'll defer to @david-koenig or @ttjsu-aws on what this setting actually does, but from my cursory check of this file, this is not an acceptable change. This seems to me like a major change to the behavior of this function.

As a general rule, don't blindly change code to make tests pass.

Reply:
I agree that this is a major change to the behavior of the function and that it should not just be merged. However, I thought it would be better to make a "suggested" change in order for the proposed assertion in line 45 of the harness to pass. In any case, is this assertion reasonable?

Reply:
Read deeper on the assertion and I think I see the problem you ran into. I'll comment over there.

Reply:
The `ciphertext_size` should be equal to the plaintext_size here as we do not include IV/authtag. I don't recommend changing this value to zero here.

Reply:
After discussing this with @ttjsu-aws offline, I am in agreement: this fundamentally changes the behavior of the function and makes the behavior no longer comply with the spec.

Reply:
After reviewing this, the `field_helper` and `field_sized` macros already add the total number of bytes needed to serialize the ciphertext to this field, so if this is _not_ set to zero, then the final `*ciphertext_size` is too large.

The main effect of this, as far as I can see, is that the output size estimate will be too large (we use this value at https://github.com/aws/aws-encryption-sdk-c/blob/e74d52d7c1595808171d79e6b6a6c2f029449724/source/session_encrypt.c#L270 where it's copied to the output size estimate). Note however that we will still succeed if the buffer size is correct, as we will only trigger a short buffer error if we did in fact run out of output buffer space.

Based on this analysis I believe this = 0 initialization is correct and corrects a minor bug, and that this exposes a gap in our unit tests.

Reply:
I looked over the code in question with @bdonlan and I concur with his assessment that changing the initialization to 0 is correct. However along with this change we should do two other things:

(1) Get rid of the comment on the previous line which is now irrelevant.
(2) Include a unit test that tests this behavior and which would have failed with the improper initialization as it was done before.

Reply:
Thanks for the response :) I will remove the comment and add a unit test that exercises this behaviour.

Reply:
I also think that the change to fix this bug and add the unit test could be separated out into a different code review rather than being mixed in with all this harness adding stuff.

Reply:
I will do it that